### PR TITLE
Code fixer improvements.

### PR DESCRIPTION
**Pull request by dragomirtitian**
Code fixer improvements.

### CODE REVIEW

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports45-no-enum-members.ts
Comment (h-joo) [inline tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports45-no-enum-members.ts:1]: Maybe rename to `...-no-computed-enum-memebers.ts`?

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports44-extract-other-to-variable.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+// @Filename: /code.ts
+////let c: string[] = [];
+////export let o = {
+////    p: Math.random() ? []: [
+////        ...c
+////    ]
+////}
+
+verify.codeFix({
```
Comment (h-joo) [inline tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports44-extract-other-to-variable.ts:15]: The two fixes here seems to be identical

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports34-inline-import-default.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/// <reference path='fourslash.ts'/>
+
+// @isolatedDeclarations: true
+// @declaration: true
+
+// @Filename: /person-code.ts
+////export type Person = { x: string; }
```
Comment (h-joo) [inline tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports34-inline-import-default.ts:7]: What happens if it's not exported? Is this worth a test case?

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports29-fn-in-object-literal.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/// <reference path='fourslash.ts'/>
+
+// @isolatedDeclarations: true
+// @declaration: true
+// fileName: code.ts
+////export const extensions = {
+////    /**
```
Comment (h-joo) [inline tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports29-fn-in-object-literal.ts:7]: Nit: Do we need this comment?

File: src/services/services.ts [hunk 4]
```diff
@@ -510,7 +510,9 @@ function addSyntheticNodes(nodes: Node[], pos: number, end: number, parent: Node
                 if (hasTabstop(parent)) {
                     continue;
                 }
-                Debug.fail(`Did not expect ${Debug.formatSyntaxKind(parent.kind)} to have an Identifier in its trivia`);
+                // Not sure why this is here. When I try to make text change that contains only a property assigment this kicks in
```
Comment (h-joo) [inline src/services/services.ts]: Can you phrase this into a TODO?

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts [hunk 5]
```diff
@@ -71,485 +126,898 @@ const declarationEmitNodeBuilderFlags = NodeBuilderFlags.MultilineObjectLiterals
...
     errorCodes,
     fixIds: [fixId],
     getCodeActions(context) {
-        const { sourceFile, span } = context;
-        const nodeWithDiag = getTokenAtPosition(sourceFile, span.start);
-
-        const changes = textChanges.ChangeTracker.with(context, t => doChange(t, context.sourceFile, context.program.getTypeChecker(), nodeWithDiag));
-        return [createCodeFixAction(fixId, changes, Diagnostics.Declaration_emit_for_this_file_requires_type_resolution_An_explicit_type_annotation_may_unblock_declaration_emit, fixId, Diagnostics.Declaration_emit_for_this_file_requires_type_resolution_An_explicit_type_annotation_may_unblock_declaration_emit)];
+        const fixes: CodeFixAction[] = [];
+        addCodeAction(addAnnotationFix, fixes, context, "full", f => f.addFullAnnotation(context.span));
```
Comment (h-joo) [inline src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:136]: Why did you need to separated these fixes?

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 6]
```diff
@@ -354,10 +354,14 @@ export function createLocalInferenceResolver({
                     const prop = objectLiteral.properties[propIndex];
 
                     if (isShorthandPropertyAssignment(prop)) {
-                        return invalid(prop);
+                        invalid(prop);
+                        inheritedObjectTypeFlags |= LocalTypeInfoFlags.Invalid;
```
Comment (h-joo) [inline src/compiler/transformers/declarations/localInferenceResolver.ts:358]: For my understanding: I guess you're trying to do the best effort here and continue after an invalid state?

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts [hunk 7]
```diff
@@ -71,485 +126,898 @@ const declarationEmitNodeBuilderFlags = NodeBuilderFlags.MultilineObjectLiterals
...
+            case SyntaxKind.PropertyDeclaration:
+            case SyntaxKind.VariableDeclaration:
+                const decl = node as ParameterDeclaration | PropertyDeclaration | VariableDeclaration;
+                return addTypeAnnotation(decl);
+            case SyntaxKind.ArrowFunction:
+            case SyntaxKind.FunctionExpression:
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.MethodDeclaration:
+            case SyntaxKind.GetAccessor:
+            case SyntaxKind.SetAccessor:
```
Comment (h-joo) [inline src/services/codefixes/fixMissingTypeAnnotationOnExports.ts]: I think  `SetAccessor` doesn't need a type annotation because it needs to be `void` and that's implied? Is this just a side effect of us not allowing any return type inference at all?

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts [hunk 8]
```diff
@@ -71,485 +126,898 @@ const declarationEmitNodeBuilderFlags = NodeBuilderFlags.MultilineObjectLiterals
...
+    cb: (fixer: Fixer) => DiagnosticOrDiagnosticAndArguments | undefined,
+) {
+    const changes = withChanges(context, typePrinter, cb);
+    if (changes.result && changes.textChanges.length) {
+        const newFix = createCodeFixAction(
+            fixName,
+            changes.textChanges,
+            changes.result,
+            fixId,
+            Diagnostics.Add_all_missing_tye_annotations,
```
Comment (h-joo) [inline src/services/codefixes/fixMissingTypeAnnotationOnExports.ts]: Is it a typo here or/and somewhere else as well? `tye_annotations`

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts [hunk 9]
```diff
@@ -71,485 +126,898 @@ const declarationEmitNodeBuilderFlags = NodeBuilderFlags.MultilineObjectLiterals
...
+        }
+    }
+
+    function handleClassDeclaration(classDecl: ClassDeclaration): DiagnosticAndArguments | undefined {
+        const extendsHeritage = classDecl.heritageClauses?.find(p => p.token === SyntaxKind.ExtendsKeyword);
+        const heritageExpression = extendsHeritage?.types[0];
+        if (!heritageExpression) {
+            return undefined;
+        }
+        // if (heritageExpression.kind !== SyntaxKind.ExpressionWithTypeArguments) {
```
Comment (h-joo) [inline src/services/codefixes/fixMissingTypeAnnotationOnExports.ts]: Can you remove this comment?

File: src/compiler/diagnosticMessages.json [hunk 10]
```diff
@@ -6992,6 +6992,42 @@
         "category": "Message",
         "code": 90060
     },
+    "Add annotation of type '{0}'": {
```
Comment (h-joo) [inline src/compiler/diagnosticMessages.json:6995]: These look great :)

I cannot find the usages, will these be standalone separate diagnostics, or do these fall under a sub-diagnostic i.e. `Diagnostic.relatedInformation` ?

File: external-declarations/src/code-mod/tsc-test-fixer/test-fixer.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    
+    try {
+        return await fixTestFiles(toBeCompiled, settings);
+    }
+    catch (e) {
+        return e as Error;
+    }
+}
+
+export async function fixTestFiles(toBeCompiled: TestFile[], settings: ts.CompilerOptions) {
```
Comment (h-joo) [inline external-declarations/src/code-mod/tsc-test-fixer/test-fixer.ts]: Does this need to be public?

File: external-declarations/src/test-runner/utils.ts [hunk 12]
```diff
@@ -62,23 +77,21 @@ export function runTypeScript(caseData: TestCaseParser.TestCaseContent, settings
             return [{
                 content: declaration?.text ?? "",
                 fileName: declarationFile,
-            }
-            // , {
-            //     content: declarationMap?.text ?? "",
-            //     fileName: declarationMapFile,
-            // }
+            }// , {
```
Comment (h-joo) [inline external-declarations/src/test-runner/utils.ts]: Lets remove this trailing comment.

File: external-declarations/src/code-mod/fixer/watch.ts
Comment (h-joo) [inline external-declarations/src/code-mod/fixer/watch.ts:1]: One question - do you intend to ship the interactive fixer in mainstream TS?

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -7227,12 +7227,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 anyType : getNonMissingTypeOfSymbol(propertySymbol);
             const saveEnclosingDeclaration = context.enclosingDeclaration;
             context.enclosingDeclaration = undefined;
-            if (context.tracker.canTrackSymbol && isLateBoundName(propertySymbol.escapedName)) {
-                if (propertySymbol.declarations) {
-                    const decl = first(propertySymbol.declarations);
+            const decl = propertySymbol.declarations && first(propertySymbol.declarations);
+
```
Comment (h-joo) [inline src/compiler/checker.ts:7231]: I don't understand what this code(the changes) is doing, can you explain with an example code snippet?

**Review by h-joo — APPROVED** [review]
LGTM with some nit and questions. Thank you for the changes!

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports34-inline-import-default.ts (hunk 2 above)
Comment (dragomirtitian) [reply tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports34-inline-import-default.ts:7]: I'll investigate. I think it might be a TS error (using a private name). But I don't think we test for it.

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts (hunk 5 above)
Comment (dragomirtitian) [reply src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:136]: They differ in the suggested type and in the position the type is placed. 

The position:
- `addFullAnnotation` - adds type to the parent declaration (variable, property etc)
- `addInlineAnnotation` - adds the type as an assertion on the expression that can't be typed in isolated declarations (this is very useful if the expression is in a large context where only a small part is creating a problem, such as a large object literal)

The type:
- `full` - TS generated type
- `relative` - an heuristic type printer that uses `typeof` extensively.

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 6 above)
Comment (dragomirtitian) [reply src/compiler/transformers/declarations/localInferenceResolver.ts:358]: Yes. exactly. This was already the behavior in most cases. It was not for shorthand property assessments which meant that only errors up to the first shorthand property assignment were reported in object literals. Which seems inconsistent with how TS usually reports errors.

File: src/compiler/checker.ts (hunk 13 above)
Comment (dragomirtitian) [reply src/compiler/checker.ts:7231]: I wanted to be able to force the compiler to preserve the computed property expressions in types instead of resolving them, if the computed property is an accessible entity name (identifier, or identifier chain):

So for example for: 

```ts
const A = 1;
const E = { A: 2 } as const
let o = { [A]: 1, [E.A]: 2 }
```
The checker would always produce as the type of `o`: `{ [1]: number, [2]: number }` which is correct but not a type a dev would ever write. With this change, if `A` and `E.A` are accessible in the current scope (either are already in the scope or can be imported), they will be preserved and we get `{ [A]: number, [E.A]: number }`

File: external-declarations/src/code-mod/fixer/watch.ts
Comment (dragomirtitian) [reply external-declarations/src/code-mod/fixer/watch.ts:1]: Probably not. A version of it will be in the tests probably, but I expect to reimplement it in a standalone project and open source that. But none of this has been set in stone, it's just my current thinking.

File: external-declarations/src/test-runner/utils.ts (hunk 12 above)
Comment (dragomirtitian) [reply external-declarations/src/test-runner/utils.ts]: Older code, will come back once we implement source maps. Removed for now.

File: external-declarations/src/code-mod/tsc-test-fixer/test-fixer.ts (hunk 11 above)
Comment (dragomirtitian) [reply external-declarations/src/code-mod/tsc-test-fixer/test-fixer.ts]: Nope. Removed export.

File: src/compiler/diagnosticMessages.json (hunk 10 above)
Comment (dragomirtitian) [reply src/compiler/diagnosticMessages.json:6995]: They are used as the fixer descriptions in `src\services\codefixes\fixMissingTypeAnnotationOnExports.ts`.

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports29-fn-in-object-literal.ts (hunk 3 above)
Comment (dragomirtitian) [reply tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports29-fn-in-object-literal.ts:7]: I want to make sure the JS doc comment is preserved in the same position. With the previous way of adding the annotation (using replaceNode), comments were duplicated. I want to make sure we do not regress this.

File: tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports44-extract-other-to-variable.ts (hunk 1 above)
Comment (dragomirtitian) [reply tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports44-extract-other-to-variable.ts:15]: Not sure what you mean. The two fixes work in tandem here. The first will extract the property value to a variable and then the second one annotates the variable.

**Review by h-joo — APPROVED** [review]

File: src/services/codefixes/fixMissingTypeAnnotationOnExports.ts (hunk 7 above)
Comment (dragomirtitian) [reply src/services/codefixes/fixMissingTypeAnnotationOnExports.ts]: You are right. Removed
